airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/StreamManager.kt

```diff
@@ -0,0 +1,219 @@
+    fun markFailed(causedBy: Exception)
+
+    /**
+     * Suspend until the stream completes successfully. Returns false if the stream is
+     * killed/failed.
+     */
+    suspend fun awaitStreamCompletedSuccessfully(): Boolean
+
+    /** Suspend until the stream completes, returning the result. */
+    suspend fun awaitStreamResult(): StreamResult
+
+    /** True if the stream has not completed. */
+    fun isActive(): Boolean
+}
+
+class DefaultStreamManager(
+    val stream: DestinationStream,
+) : StreamManager {
+    private val streamResult = CompletableDeferred<StreamResult>()
+
+    private val log = KotlinLogging.logger {}
+
+    private val recordCount = AtomicLong(0)
+    private val lastCheckpoint = AtomicLong(0L)
+
+    private val readIsClosed = AtomicBoolean(false)
+
+    private val rangesState: ConcurrentHashMap<Batch.State, RangeSet<Long>> = ConcurrentHashMap()
+
+    init {
+        Batch.State.entries.forEach { rangesState[it] = TreeRangeSet.create() }
+    }
+
+    override fun countRecordIn(): Long {
+        if (readIsClosed.get()) {
+            throw IllegalStateException("Stream is closed for reading")
+        }
+
+        return recordCount.getAndIncrement()
+    }
+
+    override fun recordCount(): Long {
+        return recordCount.get()
+    }
+
+    override fun countEndOfStream(): Long {
+        if (readIsClosed.getAndSet(true)) {
+            throw IllegalStateException("Stream is closed for reading")
+        }
+
+        return recordCount.get()
+    }
+
+    override fun closedForReading(): Boolean {
+        return readIsClosed.get()
+    }
+
+    override fun markCheckpoint(): Pair<Long, Long> {
+        val index = recordCount.get()
+        val lastCheckpoint = lastCheckpoint.getAndSet(index)
+        return Pair(index, index - lastCheckpoint)
+    }
+
+    override fun <B : Batch> updateBatchState(batch: BatchEnvelope<B>) {
+        val stateRanges =
+            rangesState[batch.batch.state]
+                ?: throw IllegalArgumentException("Invalid batch state: ${batch.batch.state}")
+
+        // Force the ranges to overlap at their endpoints, in order to work around
+        // the behavior of `.encloses`, which otherwise would not consider adjacent ranges as
+        // contiguous.
+        // This ensures that a state message received at eg, index 10 (after messages 0..9 have
+        // been received), will pass `{'[0..5]','[6..9]'}.encloses('[0..10)')`.
+        val expanded =
+            batch.ranges.asRanges().map { it.span(Range.singleton(it.upperEndpoint() + 1)) }
+
+        stateRanges.addAll(expanded)
+        log.info { "Updated ranges for $stream[${batch.batch.state}]: $stateRanges" }
+    }
+
+    /** True if all records in `[0, index)` have reached the given state. */
+    private fun isProcessingCompleteForState(index: Long, state: Batch.State): Boolean {
+        val completeRanges = rangesState[state]!!
+        println("completeRanges: $completeRanges")
```

Comment:
nit: don't forget to remove these printlns

Reply:
Thanks. Caught those locally. I'm going to tattoo `git grep println(` on my right hand.

Reply:
And `just use the debugger` on my left hand.

Reply:
println debugging is arguably superior to debugger in some cases :shrug: (_especially_ when you have tests that say things like "100 times: incrementRecordCount()" or whatever)